Rubberduck.Refactorings/Rename/RenameBaseRefactoring.cs

```diff
@@ -0,0 +1,308 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.VBEditor.ComManagement;
+using Rubberduck.VBEditor.SafeComWrappers;
+
+namespace Rubberduck.Refactorings.Rename
+{
+    public class RenameBaseRefactoring : BaseRefactoringWithSuspensionBase<RenameModel>
+    {
+        private const string AppendUnderscoreFormat = "{0}_";
+        private const string PrependUnderscoreFormat = "_{0}";
+
+        private readonly IDeclarationFinderProvider _declarationFinderProvider;
+        private readonly IProjectsProvider _projectsProvider;
+        private readonly IDictionary<DeclarationType, Action<RenameModel, IRewriteSession>> _renameActions;
+
+        public RenameBaseRefactoring(
+            IDeclarationFinderProvider declarationFinderProvider,
+            IProjectsProvider projectsProvider,
+            IParseManager parserManager,
+            IRewritingManager rewritingManager)
+            : base(parserManager, rewritingManager)
+        {
+            _declarationFinderProvider = declarationFinderProvider;
+            _projectsProvider = projectsProvider;
+
+            _renameActions = new Dictionary<DeclarationType, Action<RenameModel, IRewriteSession>>
+            {
+                {DeclarationType.Member, RenameMember},
+                {DeclarationType.Parameter, RenameParameter},
+                {DeclarationType.Event, RenameEvent},
+                {DeclarationType.Variable, RenameVariable},
+                {DeclarationType.Module, RenameModule},
+                {DeclarationType.Project, RenameProject}
+            };
+        }
+
+        protected override bool RequiresSuspension(RenameModel model)
+        {
+            //The parser needs to be suspended during the refactoring of a component because the VBE API object rename causes a separate reparse. 
+            var targetType = model.Target.DeclarationType;
+            return targetType.HasFlag(DeclarationType.Module)
+                   || targetType.HasFlag(DeclarationType.Project);
+        }
+
+        protected override void Refactor(RenameModel model, IRewriteSession rewriteSession)
+        {
+            Debug.Assert(!model.NewName.Equals(model.Target.IdentifierName, StringComparison.InvariantCultureIgnoreCase),
+                $"input validation fail: New Name equals Original Name ({model.Target.IdentifierName})");
+
```

Comment:
I've always thought that attempting to rename something to the same identifier with different casing should be allowed or handled as a no-op rather than an invalid request.  The ` Debug.Assert`  seems a bit misleading since the assert's condition, even if `true`, still allows execution in Release mode.  Full Disclosure: I'm roughly 99% sure I put this `Debug.Assert` here.

Now that this code has been moved and can be invoked free of the Rename UI's _defenses_, I think we need to decide if re-casing is allowed.  FWIW: My vote is to allow re-casing.  e.g, user wants to make his code conform to a particular casing style (like capitalizing the first letter of a subroutine name) -  seems like something RD should support.

Reply:
We couldn't *do* re-casing even if we allowed it. It's literally the same identical identifier, the VBE will simply ignore our rename operation if we let it through; "re-casing" isn't something that can be done in a case-insensitive language. Do "re-case", you need to reuse an existing name and declare a new identifier with it (or alter an existing declaration for it), using the new case - that will "re-case" every instance of that identifier.

I agree with the `Debug.Assert` being dubious, but there is no way RD can support re-casing via refactor/rename.

Reply:
Regarding the point about replicating checks, the idea is that input validation is _not_ a responsibility of the `IRefactorinAction`.

Reply:
Yeah - on further thought, I agree...it has to be that way.